wire-ios-data-model/Source/Model/UserClient/E2eIdentityProvider.swift

```diff
@@ -17,45 +17,179 @@
+        certificateStatus: E2EIdentityCertificateStatus,
         serialNumber: String
     ) {
-        self.certificateDetails = certificateDetails
+        self.details = certificateDetails
+        self.mlsThumbprint = mlsThumbprint
+        self.notValidBefore = notValidBefore
         self.expiryDate = expiryDate
-        self.certificateStatus = certificateStatus
+        self.status = certificateStatus
         self.serialNumber = serialNumber
     }
+
+    public init(
+           certificate: Certificate,
+           certificateDetails: String,
+           certificateStatus: E2EIdentityCertificateStatus,
+           mlsThumbprint: String
+       ) {
+           self.details = certificateDetails
+           self.notValidBefore = certificate.notValidBefore
+           self.expiryDate = certificate.notValidAfter
+           self.serialNumber = certificate.serialNumber.description.replacingOccurrences(of: ":", with: "")
+           self.status = certificateStatus
+           self.mlsThumbprint = mlsThumbprint
+       }
 }
 
 public protocol E2eIdentityProviding {
-    var isE2EIdentityEnabled: Bool { get }
-    func fetchCertificate() async throws -> E2eIdentityCertificate
+    func isE2EIdentityEnabled() async throws -> Bool
+    func fetchCertificates(clientIds: [Data]) async throws -> [E2eIdentityCertificate]
+    func fetchCertificates(userIds: [String]) async throws -> [String: [E2eIdentityCertificate]]
+    func shouldCertificateBeUpdated(for certificate: E2eIdentityCertificate) -> Bool
 }
 
 enum E2eIdentityCertificateError: Error {
     case badCertificate
 }
 
 public final class E2eIdentityProvider: E2eIdentityProviding {
-    public var isE2EIdentityEnabled: Bool
 
-    // TODO: Change this upon implementing business logic
-    public init(isE2EIdentityEnabled: Bool = false) {
-        self.isE2EIdentityEnabled = isE2EIdentityEnabled
+    // Grace period fromt he E2ei config settings
+    private let gracePeriod: Double
+    private let coreCryptoProvider: CoreCryptoProviderProtocol
+    private let conversationId: Data
+
+    public init(
+        gracePeriod: Double,
+        coreCryptoProvider: CoreCryptoProviderProtocol,
+        conversationId: Data // make sure to run on SyncContext
+    ) {
+        // TODO: precondition(zm_isSyncContext, "MLSService should only be accessed on the sync context")
+        self.gracePeriod = gracePeriod
+        self.coreCryptoProvider = coreCryptoProvider
+        self.conversationId = conversationId
+    }
+
+    var coreCrypto: SafeCoreCryptoProtocol {
+        get async throws {
+           try await coreCryptoProvider.coreCrypto(requireMLS: true)
+        }
+    }
+    public func isE2EIdentityEnabled() async throws -> Bool {
+        try await coreCrypto.perform {
+             try await $0.e2eiIsEnabled(ciphersuite: CiphersuiteName.default.rawValue)
+        }
+    }
+
+    public func fetchCertificates(clientIds: [Data]) async throws -> [E2eIdentityCertificate] {
```

Comment:
Shouldn't we be using our `MLSClientID` type here?

```suggestion
    public func fetchCertificates(clientIds: [MLSClientID]) async throws -> [E2eIdentityCertificate] {
```

Reply:
Yeah. Makes sense